internal/pkg/pkg.go

```diff
@@ -145,14 +144,63 @@ func (p *Pkg) RelativePathTo(ancestorPkg *Pkg) (string, error) {
 // TODO: This does not support symlinks, so we need to figure out how
 // we should support that with kpt.
 func (p *Pkg) DirectSubpackages() ([]*Pkg, error) {
+	var subPkgs []*Pkg
+
+	packagePaths, err := Subpackages(p.UniquePath.String(), All, false)
+	if err != nil {
+		return subPkgs, err
+	}
+
+	for _, subPkgPath := range packagePaths {
+		subPkg, err := New(filepath.Join(p.UniquePath.String(), subPkgPath))
+		if err != nil {
+			return subPkgs, fmt.Errorf("failed to read subpkg at path %s %w", subPkgPath, err)
+		}
+		subPkgs = append(subPkgs, subPkg)
+	}
+
+	sort.Slice(subPkgs, func(i, j int) bool {
+		return subPkgs[i].DisplayPath < subPkgs[j].DisplayPath
+	})
+	return subPkgs, nil
+}
+
+// SubpackageMatcher is type for specifying the types of subpackages which
+// should be included when listing them.
+type SubpackageMatcher string
+
+const (
+	// All means all types of subpackages will be returned.
+	All SubpackageMatcher = "ALL"
+	// Local means only local subpackages will be returned.
+	Local SubpackageMatcher = "LOCAL"
+	// remote means only remote subpackages will be returned.
+	Remote SubpackageMatcher = "REMOTE"
+)
+
+// Subpackages returns a slice of paths to any subpackages of the provided path.
+// The matcher parameter decides if all types of subpackages should be considered,
+// and the recursive parameter determines if only direct subpackages are
+// considered. All returned paths will be relative to the provided rootPath.
+// The top level package is not considered a subpackage. If the provided path
+// doesn't exist, an empty slice will be returned.
+// Symlinks are ignored.
+func Subpackages(rootPath string, matcher SubpackageMatcher, recursive bool) ([]string, error) {
```

Comment:
Do you want to export this function? It uses raw string for input and output instead of specific types,.

Reply:
Yeah, at least for now I need it to just accept strings. The update code needs to use this functionality on clones of git repos in temporary directories, so the `DisplayPath` functionality of package doesn't work all that well. It also involves traversing multiple directory hierarchies in parallel, so it needs to manipulate paths. 
So this isn't perfect, but I think it is a step in the right direction. This gives up one tested implementation of looking up subpackages. Will continue to see if we can rely more on the packaging abstraction.

Reply:
Add a TODO for ideal end-state?

Reply:
Added a TODO.